### ISSUE TITLE
[CI] ClassReadersTests testMultipleJarsInADir failing

### ISSUE THREAD

**Issue by cbuescher** [elastic/elasticsearch#97166]
Several other ClassReadersTests and some in NamedComponentScannerTests are failing as well.

**Build scan:**
https://gradle-enterprise.elastic.co/s/taevwfbqkt6lc/tests/:libs:elasticsearch-plugin-scanner:test/org.elasticsearch.plugin.scanner.ClassReadersTests/testMultipleJarsInADir

**Reproduction line:**
```
./gradlew ':libs:elasticsearch-plugin-scanner:test' --tests "org.elasticsearch.plugin.scanner.ClassReadersTests.testMultipleJarsInADir" -Dtests.seed=D9590B0A8ABB175F -Dtests.locale=sv -Dtests.timezone=Pacific/Rarotonga -Druntime.java=21
```

**Applicable branches:**
main

**Reproduces locally?:**
Didn't try

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.plugin.scanner.ClassReadersTests&tests.test=testMultipleJarsInADir

**Failure excerpt:**
```
java.lang.IllegalArgumentException: Unsupported class file major version 65

  at __randomizedtesting.SeedInfo.seed([D9590B0A8ABB175F:D567D812022E7F5]:0)
  at org.objectweb.asm.ClassReader.<init>(ClassReader.java:199)
  at org.objectweb.asm.ClassReader.<init>(ClassReader.java:180)
  at org.objectweb.asm.ClassReader.<init>(ClassReader.java:166)
  at org.elasticsearch.plugin.scanner.ClassReaders.lambda$classesInPath$5(ClassReaders.java:97)
  at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
  at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
  at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
  at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
  at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
  at java.util.Iterator.forEachRemaining(Iterator.java:133)
  at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1924)
  at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
  at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
  at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
  at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
  at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
  at org.elasticsearch.plugin.scanner.ClassReaders.classesInPath(ClassReaders.java:102)
  at org.elasticsearch.plugin.scanner.ClassReaders.classesInJar(ClassReaders.java:83)
  at org.elasticsearch.plugin.scanner.ClassReaders.lambda$ofPaths$1(ClassReaders.java:73)
  at java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:273)
  at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
  at java.util.Iterator.forEachRemaining(Iterator.java:133)
  at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1924)
  at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
  at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
  at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:575)
  at java.util.stream.AbstractPipeline.evaluateToArrayNode(AbstractPipeline.java:260)
  at java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:616)
  at java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:622)
  at java.util.stream.ReferencePipeline.toList(ReferencePipeline.java:627)
  at org.elasticsearch.plugin.scanner.ClassReaders.ofPaths(ClassReaders.java:77)
  at org.elasticsearch.plugin.scanner.ClassReaders.ofDirWithJars(ClassReaders.java:48)
  at org.elasticsearch.plugin.scanner.ClassReadersTests.testMultipleJarsInADir(ClassReadersTests.java:125)
  at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
  at java.lang.reflect.Method.invoke(Method.java:580)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:1583)

```

**Comment by cbuescher** [comment]
Reproduces locally for me:
```
java.lang.IllegalArgumentException: Unsupported class file major version 65
        at __randomizedtesting.SeedInfo.seed([D9590B0A8ABB175F:D567D812022E7F5]:0)
        at org.objectweb.asm.ClassReader.<init>(ClassReader.java:199)
        at org.objectweb.asm.ClassReader.<init>(ClassReader.java:180)
        at org.objectweb.asm.ClassReader.<init>(ClassReader.java:166)
        at org.elasticsearch.plugin.scanner.ClassReaders.lambda$classesInPath$5(ClassReaders.java:97)
```

**Comment by thecoop** [comment]
This is due to using jdk 21, which introduces a new class file version that, it looks like, asm does not support

**Comment by cbuescher** [comment]
Also fails on 8.9: https://gradle-enterprise.elastic.co/s/nuhscjxxmnxnw

**Comment by thecoop** [comment]
Should be fixed by #97296
